File: backend/src/main/java/com/fitnexus/backend/entity/FitnessUser.java
package com.fitnexus.backend.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Entity
@Table(name = "users")
@Data
@NoArgsConstructor
@AllArgsConstructor
public class FitnessUser {
    @Id
    @Column(name = "user_id")
    private Long id;

    @OneToOne(fetch = FetchType.LAZY, optional = false, cascade = CascadeType.ALL)
    @MapsId
    @JoinColumn(name = "user_id")
    private Users account;

    @Column(name = "first_name", length = 50)
    private String firstName;

    @Column(name = "last_name", length = 50)
    private String lastName;

    @Column(name = "weight")
    private Double weight;

    @Column(name = "height")
    private Double height;

    @Column(name = "blood_group", length = 3)
    private String bloodGroup;

    @Column(name = "age")
    private Integer age;

    @Column(name = "gender", length = 20)
    private String gender;

    @Column(name = "phone", length = 15)
    private String phone;

    @Column(name = "address", length = 200)
    private String address;

    @Column(name = "city", length = 50)
    private String city;

    @Column(name = "medical_conditions", length = 500)
    private String medicalConditions;

    @Column(name = "emergency_contact", length = 100)
    private String emergencyContact;

    @Column(name = "emergency_phone", length = 15)
    private String emergencyPhone;

    /** Member profile photo (uploaded file path) — collected at registration. */
    @Column(name = "photo_path", length = 500)
    private String photoPath;

    /** Member ID proof document (e.g. Aadhaar) — uploaded file path. */
    @Column(name = "id_proof_path", length = 500)
    private String idProofPath;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_trainer_id")
    private Trainer assignedTrainer;

    @Column(name = "details_completed", nullable = false)
    private Boolean detailsCompleted = false;

    @Column(name = "registration_date")
    private LocalDateTime registrationDate;

    @Column(name = "activation_date")
    private LocalDateTime activationDate;

    @Column(name = "bio", length = 500)
    private String bio;

    @Column(name = "date_of_birth")
    private LocalDateTime dateOfBirth;

    @Column(name = "fitness_goals", length = 300)
    private String fitnessGoals;

    @Column(name = "dietary_preferences", length = 300)
    private String dietaryPreferences;

    @Column(name = "injuries_or_limitations", length = 500)
    private String injuriesOrLimitations;

    @Column(name = "body_fat")
    private Double bodyFat;

    @Column(name = "is_frozen", nullable = false, columnDefinition = "boolean default false")
    private Boolean isFrozen = false;

    @Column(name = "referred_by", length = 100)
    private String referredBy;

    @Column(name = "membership_plan", length = 20)
    private String membershipPlan = "BASIC";

    @Column(name = "membership_expiry")
    private java.time.LocalDate membershipExpiry;

    /** The configurable plan tier this member is on (source of truth for unlimited access). */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "membership_plan_id")
    private MembershipPlan membershipPlanRef;

    /** Daily check-in window assigned to this member (ignored when the plan is unlimited). */
    @Column(name = "access_start_time")
    private java.time.LocalTime accessStartTime;

    @Column(name = "access_end_time")
    private java.time.LocalTime accessEndTime;

    /** Last time we alerted staff that this member has no slot assigned (throttles the alert). */
    @Column(name = "last_slot_request_at")
    private LocalDateTime lastSlotRequestAt;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "assigned_corporate_hr_id")
    private Users assignedCorporateHr;

    @Column(name = "corporate_department", length = 100)
    private String corporateDepartment;

    @PrePersist
    protected void onCreate() {
        if (registrationDate == null) {
            registrationDate = LocalDateTime.now();
        }
    }
}

File: backend-fastapi/app/models/fitness_user.py
from datetime import date, datetime, time

from sqlalchemy import Boolean, Date, DateTime, Float, ForeignKey, Integer, String, Time
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class FitnessUser(Base):
    """Mirrors entity/FitnessUser.java — table `users` (member profile), 1:1 with user_accounts."""

    __tablename__ = "users"

    id: Mapped[int] = mapped_column(
        "user_id", ForeignKey("user_accounts.id"), primary_key=True
    )
    account: Mapped["Users"] = relationship("Users", foreign_keys=[id])

    first_name: Mapped[str | None] = mapped_column(String(50))
    last_name: Mapped[str | None] = mapped_column(String(50))
    weight: Mapped[float | None] = mapped_column(Float)
    height: Mapped[float | None] = mapped_column(Float)
    blood_group: Mapped[str | None] = mapped_column(String(3))
    age: Mapped[int | None] = mapped_column(Integer)
    gender: Mapped[str | None] = mapped_column(String(20))
    phone: Mapped[str | None] = mapped_column(String(15))
    address: Mapped[str | None] = mapped_column(String(200))
    city: Mapped[str | None] = mapped_column(String(50))
    medical_conditions: Mapped[str | None] = mapped_column(String(500))
    emergency_contact: Mapped[str | None] = mapped_column(String(100))
    emergency_phone: Mapped[str | None] = mapped_column(String(15))
    photo_path: Mapped[str | None] = mapped_column(String(500))
    id_proof_path: Mapped[str | None] = mapped_column(String(500))

    # FK to trainers table, added once the Trainer model exists (Org/People phase).
    assigned_trainer_id: Mapped[int | None] = mapped_column(Integer)

    details_completed: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    registration_date: Mapped[datetime | None] = mapped_column(DateTime)
    activation_date: Mapped[datetime | None] = mapped_column(DateTime)
    bio: Mapped[str | None] = mapped_column(String(500))
    date_of_birth: Mapped[datetime | None] = mapped_column(DateTime)
    fitness_goals: Mapped[str | None] = mapped_column(String(300))
    dietary_preferences: Mapped[str | None] = mapped_column(String(300))
    injuries_or_limitations: Mapped[str | None] = mapped_column(String(500))
    body_fat: Mapped[float | None] = mapped_column(Float)
    is_frozen: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
    referred_by: Mapped[str | None] = mapped_column(String(100))
    membership_plan: Mapped[str | None] = mapped_column(String(20), default="BASIC")
    membership_expiry: Mapped[date | None] = mapped_column(Date)

    # FK to membership_plans table, added once that model exists (Membership phase).
    membership_plan_id: Mapped[int | None] = mapped_column(Integer)

    access_start_time: Mapped[time | None] = mapped_column(Time)
    access_end_time: Mapped[time | None] = mapped_column(Time)
    last_slot_request_at: Mapped[datetime | None] = mapped_column(DateTime)

    assigned_corporate_hr_id: Mapped[int | None] = mapped_column(ForeignKey("user_accounts.id"))
    corporate_department: Mapped[str | None] = mapped_column(String(100))
